### ISSUE TITLE
Establish a worker/app communication protocol

### ISSUE THREAD

**Issue by eddieantonio** [eddieantonio/predictive-text-studio#74]
The app (UI thread) should instantiate the worker once; then, the app and the worker can communicate via `postMessage()`. But _how_ should they communicate. Here's a proposal:

Use [discriminating unions](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions) to establish messages and their payload.

We first list the different kinds of messages that can be sent:

```typescript
type Message = UploadFileMessage | CompilationCompleteMessage;
```

We will add many, _many_ more messages!

These are interfaces that both the worker and app know (see #73).

All of these interfaces have one **common property**. Let's call it `method`:

```
interface UploadFileMessage {
    method: "upload-file";

    filename: string;
    rawData: ArrayBuffer;
}

interface CompilationCompleteMessage {
    method: "compiliation-complete";

    downloadURL: string;
}
```

---

Usage:

In both the app and the worker, we send one of these messages to `postMesage()`:

```typescript
// in the app
worker.postMessage({
  method: "upload-file",
  filename: "example.xlsx",
  rawData: await file.arrayBuffer(),
});

// in the worker
self.postMessage({
   method: "compilation-complete",
   downloadURL: URL.createObjectURL(kmpFile),
});
```

Related to #73

**Comment by eddieantonio** [comment]
Hey @IanSeng, any progress on this? I'm realizing that #53 touches upon this, and we might need to make use of the work in this issue over in that PR.

If you haven't started, could you please unassign yourself, and move this to the "TODO" column?

**Comment by IanSeng** [comment]
> Hey @IanSeng, any progress on this? I'm realizing that #53 touches upon this, and we might need to make use of the work in this issue over in that PR.
> 
> If you haven't started, could you please unassign yourself, and move this to the "TODO" column?

Apologies for the delayed response, just got email notified. Since I haven't started working on this task and there is a dependency with #53, I will unassign myself for this one and pick up another.

**Comment by eddieantonio** [comment]
@jienius, I think #83 now actually covers _this_ issue instead of #53!

**Comment by eddieantonio** [comment]
I've assigned this to @jienius as #83 is essentially solving this issue! 😂

**Comment by eddieantonio** [comment]
Resolved in #94 🥳
